ocaml/xapi/xapi_psr.ml

```diff
@@ -329,61 +465,72 @@ end)
+    let self = Helpers.get_pool ~__context in
+    let assert_no_ha () =
+      let is_ha_enabled = Db.Pool.get_ha_enabled ~__context ~self in
+      if is_ha_enabled then
+        raise Api_errors.(Server_error (ha_is_enabled, []))
+    in
+    let assert_we_are_master () =
+      if
+        not
+          (Helpers.is_pool_master ~__context
+             ~host:(Helpers.get_localhost ~__context))
+      then
         raise
           Api_errors.(
-            Server_error (cannot_contact_host, [Ref.string_of offline_host]))
-    ) ;
-    let is_ha_enabled =
-      Db.Pool.get_ha_enabled ~__context ~self:(Helpers.get_pool ~__context)
+            Server_error (host_is_slave, [Pool_role.get_master_address ()]))
+    in
+    let assert_all_hosts_alive () =
+      let live_hosts = Helpers.get_live_hosts ~__context |> HostSet.of_list in
+      let all_hosts_list =
+        Xapi_pool_helpers.get_master_slaves_list ~__context
+      in
+      let all_hosts = all_hosts_list |> HostSet.of_list in
+      let offline_hosts = HostSet.diff all_hosts live_hosts in
+      ( if not (HostSet.is_empty offline_hosts) then
+          let offline_host = HostSet.min_elt offline_hosts in
+          raise
+            Api_errors.(
+              Server_error (cannot_contact_host, [Ref.string_of offline_host]))
+      ) ;
+      all_hosts_list
     in
-    if is_ha_enabled then raise Api_errors.(Server_error (ha_is_enabled, [])) ;
-    if Helpers.rolling_upgrade_in_progress ~__context then
-      raise Api_errors.(Server_error (not_supported_during_upgrade, [])) ;
-    let module PSR = Make (Impl (struct let __context = __context end)) in
-    let[@warning "-8"] (master :: members) = all_hosts_list in
-    let r =
-      PSR.start
-        (Xapi_globs.pool_secret (), Ptoken.gen_token ())
-        ~master ~members
+    let assert_no_rpu () =
+      if Helpers.rolling_upgrade_in_progress ~__context then
+        raise Api_errors.(Server_error (not_supported_during_upgrade, []))
     in
-    match r with
-    | Ok () ->
-        D.info "Xapi_psr.start: pool secret rotation successful"
-    | Error e ->
-        let err_msg = user_facing_error_message ~__context e in
-        D.error "PSR failed: %s" err_msg ;
-        raise Api_errors.(Server_error (internal_error, [err_msg]))
-  in
-  if Mutex.try_lock m then (
-    try f () ; Mutex.unlock m with e -> Mutex.unlock m ; raise e
-  ) else
-    raise
-      Api_errors.(
-        Server_error (internal_error, ["pool secret rotation already running"]))
+    with_lock (fun () ->
+        Pool_features.assert_enabled ~__context ~f:Features.Pool_secret_rotation ;
+        assert_we_are_master () ;
+        Assert.master_state_valid () ;
+        let[@warning "-8"] (master :: members) = assert_all_hosts_alive () in
+        assert_no_ha () ;
+        assert_no_rpu () ;
+        let module PSR = Make (Impl (struct let __context = __context end)) in
+        Db.Pool.set_is_psr_pending ~__context ~self ~value:true ;
```

Comment:
I think this should be set before asserting ha is not enabled to avoid race conditions, however small they might be. (and the flag reset if the assertion fails)